### ISSUE TITLE
Gatsby re-renders all components when clicking back/forward buttons (only in production build)

### ISSUE THREAD

**Issue by LordCarlosMP** [gatsbyjs/gatsby#38275]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Gatsby re-renders all components no matter their location in the tree when clicking back/forward buttons in the browser.
This happens only in production builds created with gatsby build, in development builds with gatsby develop it does not happen.

This bug makes navigating forward / backwards (even without changing URL, only using HistoryAPI) very inefficient, and also voids any cached state you may have in a component that you are passing down with a Context, causing divergent behaviour in development/production environments).

### Reproduction Link

https://github.com/LordCarlosMP/GatsbyBackForward

### Steps to Reproduce

1. Create gatsby project from scratch: gatsby new (or clone the reproduction repository)
2. in src/pages/index.js, write this code:
`import * as React from 'react';

const IndexPage = () => {
  console.log('Re-rendering');
  return <button onClick={() => {
    window.history.replaceState({state: "newBrowserHistoryState"}, '', '');
  }}>
    Button
  </button>;
};

export default IndexPage;

export const Head = () => <title>Home Page</title>;
`

3. run gatsby develop (in development works. it is just to see the difference from production)
4. Click the button a couple times to create some history
5. Navigate Backwards / Forwards with browser buttons. You will see that the "Re-rendering" message appears only once, at the initial render
6. Now, create and serve a production build: gatsby build && gatsby serve
7. Click the button a couple times to create some history
8. Navigate Backwards / Forwards with browser buttons. You will get the "Re-rendering" text on each back/forward click

You can make further testing by adding child components and see how they also re-render, even without any props or state

### Expected Result

Production and development environments should behave the same way. No re-renders should occur on clicking back/forward on production

### Actual Result

In production, all components get re-rendered when clicking

### Environment

```shell
System:
    OS: Linux 6.3 Garuda Linux
    CPU: (32) x64 AMD Ryzen 9 5950X 16-Core Processor
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 20.3.0 - /usr/bin/node
    npm: 8.19.2 - /usr/bin/npm
  Languages:
    Python: 3.11.3 - /usr/bin/python
  npmPackages:
    gatsby: ^5.11.0 => 5.11.0
  npmGlobalPackages:
    gatsby-cli: 5.2.0
    gatsby: 4.8.0
```

**Comment by LekoArts** [comment]
Hi!

For client navigation you'll need to use the Gatsby Link API: https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-link/

You'll need to use the `<Link>` component or the `navigate` function, not `window.history.replaceState`.
